src/dag/key.rs

```diff
@@ -4,43 +4,47 @@ use std::str;
 // KVKey is the key we use to store our dag data in the underlying
 // kvstore.
 #[derive(Debug, PartialEq, Eq)]
-pub enum Key {
-    ChunkData(String),
-    ChunkRefs(String),
-    Head(String),
+pub enum Key<'a> {
+    ChunkData(&'a str),
+    ChunkMeta(&'a str),
+    Head(&'a str),
 }
 
-impl fmt::Display for Key {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            Key::ChunkData(hash) => write!(f, "c/{}/d", hash),
-            Key::ChunkRefs(hash) => write!(f, "c/{}/r", hash),
-            Key::Head(name) => write!(f, "h/{}", name),
+type ParseError = ();
+
+// TODO: It would be cool to make this implement FromString trait,
+// as that is the convention, and then "foo".parse() would work.
+// But I got lost in lifetime goop.
+impl<'a> Key<'_> {
+    pub fn parse<'b>(s: &'b str) -> Result<Key<'b>, ParseError> {
+        let mut parts = s.split::<'b>("/");
+        let prefix: &str = parts.next().ok_or(())?;
+        let content = parts.next().ok_or(())?;
+        match prefix {
+            "c" => {
+                let suffix = parts.next().ok_or(())?;
+                if parts.next().is_some() {
+                    return Err(());
+                }
+                match suffix {
+                    "d" => Ok(Key::ChunkData(content)),
+                    "m" => Ok(Key::ChunkMeta(content)),
+                    _ => Err(()),
+                }
+            },
+            "h" => Ok(Key::Head(content)),
+            _ => Err(()),
         }
     }
 }
 
-type ParseError = ();
-
-impl str::FromStr for Key {
-    type Err = ParseError;
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let parts = s.split("/").collect::<Vec<&str>>();
-        if parts.len() >= 2 {
-            let name = parts[1].into();
-            if parts[0] == "c" {
-                if parts.len() == 3 {
-                    if parts[2] == "d" {
-                        return Ok(Key::ChunkData(name));
-                    } else if parts[2] == "r" {
-                        return Ok(Key::ChunkRefs(name));
-                    }
-                }
-            } else if parts[0] == "h" {
-                return Ok(Key::Head(name));
-            }
+impl<'a> fmt::Display for Key<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Key::ChunkData(hash) => write!(f, "c/{}/d", hash),
```

Comment:
Not specific to this commit, but I just wanted to register that I'm a little iffy on these one-letter prefixes. Just something to think about as we keep building out.